### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by ewjoachim**
Ping @Evelf 

We now have 100% coverage including:
- 98% coverage from unit and integration (postgres) test
- 93% coverage from acceptance test

That's what I'd call a good start !

### CODE REVIEW

**Comment by ewjoachim** [comment]
We now have 100% coverage including:
- 98% coverage from unit and integration (postgres) test
- 93% coverage from acceptance test

That's what I'd call a good start !

File: tests/integration/test_postgres.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    result = get_all("queues", "queue_name")
+    assert result == [{"queue_name": "marsupilami"}]
+
+
+def test_listen_queue(db):
+    with db.cursor() as cursor:
+        postgres.listen_queue(curs=cursor, queue="bla")
+        db.commit()
+        cursor.execute("""select * from pg_listening_channels() """)
+        assert cursor.fetchone()[0] == "queue#bla"
```
Comment (ewjoachim) [inline tests/integration/test_postgres.py]: Ping @PierreDucroquet :tada:

Comment (PierreDucroquet) [reply tests/integration/test_postgres.py]: As you wish… mais c'est un peu jouer avec le feu, non ?
```
import uuid
queue = str(uuid.uuid4())
postgres.listen_queue(curs=cursor, queue=queue)
db.commit()
cursor.execute("""select * from pg_listening_channels() WHERE pg_listening_channels = %s""", (queue,))
assert cursor.fetchone()[0] == ("queue#%s" % queue)
```

=> t'es certain de pas te manger les pieds dans le tapis le jour où ton test unitaire évolue et où tu testes plusieurs choses en parallèle… Et ça coûte pas plus cher, non ?

Comment (ewjoachim) [reply tests/integration/test_postgres.py]: (let's keep it in :gb: , we're on an opensource project :) )

Play with fire, I'd say is a small overstatement. according to the doc, `pg_listening_channels` returns listeners for the current session. I'd tend to say that even if we were to run tests in parallel, on the same db, we'd need to use the same psycopg2 connection object to really start having problems.

That said, if you're more confortable that way, I have nothing against implementing that.

Comment (ewjoachim) [reply tests/integration/test_postgres.py]: (done)

Comment (PierreDucroquet) [reply tests/integration/test_postgres.py]: Sorry, forgot this one was public. As I said : "as you wish".
But "bla" is not a good english-looking name, if I may... :)

Comment (ewjoachim) [reply tests/integration/test_postgres.py]: You may. Foo bar baz is overrated sometimes. And I'm sure if we used other languages idle talk, especially non-ascii ones, we'd find interesting bug :D

File: cabbage/task_worker.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+def one_loop(task_manager: tasks.TaskManager, queue: str, curs: Any, timeout) -> None:
+    process_tasks(task_manager=task_manager, queue=queue, curs=curs)
+    logger.debug("waiting")
+    select.select([curs.connection], [], [], timeout)
+
+
+def process_tasks(task_manager: tasks.TaskManager, queue: str, curs: Any) -> None:
+
+    for task_row in postgres.get_tasks(cursor=curs, queue=queue):  # pragma: no branch
+        if task_row["id"] is None:
```
Comment (mgu) [inline cabbage/task_worker.py]: please move this code to get_tasks and add a comment

File: cabbage/postgres.py [hunk 3]
```diff
@@ -1,39 +1,85 @@
...
+        row = cursor.fetchone()
+
+    if not row:
+        raise exceptions.QueueNotFound(queue)
+
     conn.commit()
+    return row[0]
+
+
+def get_tasks(cursor: Any, queue: str) -> Iterator[Dict[str, Union[str, int]]]:
```
Comment (mgu) [inline cabbage/postgres.py]: return a namedtupled instead of dict : it will be easier to handle typing checks

File: cabbage/task_worker.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    description = f"{task.queue}.{task.name}.{pk}({kwargs})"
+    logger.info(f"Start - {description}")
+    try:
+        task_run.run(**kwargs)
+    except Exception as e:
+        logger.exception(f"Error - {description}")
+        raise exceptions.TaskError() from e
+    else:
+        logger.info(f"Succesfully ran task {task.name} with args {kwargs}")
```
Comment (mgu) [inline cabbage/task_worker.py]:
```suggestion
        logger.info(f"Success - {description}")
```

File: cabbage/task_worker.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        task = task_manager.tasks[task_name]
+    except KeyError:
+        raise exceptions.TaskNotFound(task_row)
+
+    pk = task_row["id"]
+
+    task_run = tasks.TaskRun(task=task, id=pk, lock=task_row["targeted_object"])
+    kwargs = task_row["args"]
+
+    description = f"{task.queue}.{task.name}.{pk}({kwargs})"
```
Comment (mgu) [inline cabbage/task_worker.py]: could we use a generic message and the variables in `extra=` ?

File: cabbage/tasks.py [hunk 6]
```diff
@@ -1,40 +1,49 @@
...
+        if not self.name:
+            self.name = func.__name__
         self.manager.register(self)
         return self
 
-    def defer(self, lock=None, **kwargs):
+    def defer(self, lock: str = None, **kwargs: types.JSONValue) -> None:
         lock = lock or f"{uuid.uuid4()}"
         logger.info(
             f"Scheduling task {self.name} in queue {self.queue} with args {kwargs}"
```
Comment (mgu) [inline cabbage/tasks.py]: generic message + `extra` ?

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+@pytest.fixture(scope="session")
+def setup_db():
+
+    with closing(psycopg2.connect("", dbname="postgres")) as connection:
+        connection.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        with connection.cursor() as cursor:
+            _execute(cursor, "DROP DATABASE IF EXISTS {}", "cabbage_test_template")
+            _execute(cursor, "CREATE DATABASE {}", "cabbage_test_template")
+
+    with psycopg2.connect("", dbname="cabbage_test_template") as connection:
```
Comment (mgu) [inline tests/conftest.py]:
```suggestion
    with closing(psycopg2.connect("", dbname="cabbage_test_template")) as connection:
```

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        with connection.cursor() as cursor:
+            with open("init.sql") as migrations:
+                cursor.execute(migrations.read())
+        connection.commit()
+        connection.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        yield connection
+
+    connection.close()
```
Comment (mgu) [inline tests/conftest.py]:
```suggestion
```

File: cabbage/task_worker.py (hunk 5 above)
Comment (ewjoachim) [reply cabbage/task_worker.py]: We'll do that in a dedicated PR (issue #4)

File: cabbage/tasks.py (hunk 6 above)
Comment (ewjoachim) [reply cabbage/tasks.py]: #4

File: cabbage/postgres.py [hunk 9]
```diff
@@ -1,39 +1,102 @@
...
+
 import psycopg2
+from psycopg2 import extras, sql
+from psycopg2.extras import RealDictCursor
+
+from cabbage import exceptions, types
+
 
-from psycopg2 import sql
+class TaskRow(NamedTuple):
```
Comment (mgu) [inline cabbage/postgres.py]: :heart:

**Pull request merged**